RandBLAS/trig_skops.hh

```diff
@@ -0,0 +1,538 @@
+                  //NOTE: I am also going to assume that the user is kind enough to
+                  // populate this fully with zeros
+    //NOTE: We could (should?) explore the possibility of managing the buffer by ourselves
+    int64_t workspace_ld;
+    bool left;
+
+    // Trivial constructor
+    HadamardMatrixCouple(
+        bool left,
+        blas::Layout layout,
+        int64_t rows,
+        int64_t cols,
+        T* A,
+        int64_t lda,
+        T* workspace
+    ) : left(left), layout(layout), n_rows(rows), n_cols(cols), A(A), lda(lda), workspace(workspace) {};
+};
+
+/*
+ * A `pure` description of the \PiRHT as a «data-oblvious» transform
+ */
+template <typename T, SignedInteger sint_t = int64_t, typename state_t = RNGState<DefaultRNG>>
+struct HadamardMixingOp{
+    sint_t* diag_scale = nullptr;
+    sint_t* selected_idxs = nullptr;
+    const state_t seed_state;
+    int64_t dim; // The dimension to be sketched
+    int64_t d; // The number of cols/rows to be permuted at the end of the operation
+    bool own_memory = false; // Do we manage our own memory?
+
+    // Constructor, we manage the memory
+    HadamardMixingOp(int64_t dim,
+                     int64_t d,
+                     state_t seed_state
+                    ) : dim(dim), d(d), seed_state(seed_state), own_memory(true) {};
+
+
+    // Constructor, user-managed memory
+    HadamardMixingOp(int64_t dim,
+                     int64_t d,
+                     sint_t* diag_scale,
+                     sint_t* selected_idxs
+                    ) : dim(dim), d(d), diag_scale(diag_scale), selected_idxs(selected_idxs), seed_state(seed_state), own_memory(false) {};
+
+    bool filled() {
+        if(this -> diag_scale == nullptr \
+        && this -> selected_idxs == nullptr)
+            return false;
+        else
+            return true;
+    }
+
+    // Destructor
+    ~HadamardMixingOp() {
+        if (own_memory) {
+            if (diag_scale != nullptr) delete [] diag_scale;
+            if (selected_idxs != nullptr) delete [] selected_idxs;
+        }
+    }
+
+    private:
+};
+
+/*
+* Populates a `HadamardMixingOp`: `selected_idxs` and `diag_scale`
+*/
+template <typename T, typename state_t = RNGState<DefaultRNG>, SignedInteger sint_t = int64_t>
+state_t fill_hadamard(
+                    HadamardMixingOp<T, sint_t> &hmo
+                  ) {
+    if(!hmo.filled()) {
+        hmo.diag_scale = new sint_t[hmo.dim];
+
+        hmo.selected_idxs = new sint_t[hmo.d];
+
+        auto [ctr, key] = hmo.seed_state;
+
+        // Populating `diag`
+        auto next_state = generate_rademacher_vector_r123(hmo.diag_scale, hmo.dim, hmo.seed_state);
+
+        // Populating `selected_idxs`
+        next_state = repeated_fisher_yates<sint_t>(
+            hmo.d,
+            hmo.dim,
+            1,
+            hmo.selected_idxs,
+            next_state
+        );
+
+        return next_state;
+    }
+}
```

Comment:
fill functions that accept a datastructure, like a SketchingOperator, should not return a next state. The next state is computed at construction time with a helper function that performs _no random sampling_. Example:
https://github.com/BallisticLA/RandBLAS/blob/fd180c5743757936fdfa5be8b0fecc62d66b55c4/RandBLAS/sparse_skops.hh#L243-L260

and

https://github.com/BallisticLA/RandBLAS/blob/fd180c5743757936fdfa5be8b0fecc62d66b55c4/RandBLAS/sparse_skops.hh#L376-L384